v-next/core/test/internal/global-options.ts

```diff
@@ -240,7 +242,202 @@ describe("Global Options", () => {
     });
   });
 
-  describe.todo("resolveGlobalOptions", () => {
-    // TODO: Implement tests.
+  describe("resolveGlobalOptions", () => {
+    const { setEnvVar } = createTestEnvManager();
+
+    it("should resolve to the default values when no options are provided", () => {
+      const globalOptionsMap = buildGlobalOptionsMap([
+        {
+          id: "plugin1",
+          globalOptions: [
+            buildGlobalOptionDefinition({
+              name: "param1",
+              description: "param1 description",
+              parameterType: ParameterType.BOOLEAN,
+              defaultValue: true,
+            }),
+            buildGlobalOptionDefinition({
+              name: "param2",
+              description: "param2 description",
+              defaultValue: "default",
+            }),
+          ],
+        },
+      ]);
+
+      const globalOptions = resolveGlobalOptions({}, globalOptionsMap);
+
+      assert.deepEqual(globalOptions, {
+        param1: true,
+        param2: "default",
+      });
+    });
+
+    it("should resolve to the user provided options and env variables", () => {
+      const globalOptionsMap = buildGlobalOptionsMap([
+        {
+          id: "plugin1",
+          globalOptions: [
+            buildGlobalOptionDefinition({
+              name: "param1",
+              description: "param1 description",
+              parameterType: ParameterType.BOOLEAN,
+              defaultValue: true,
+            }),
+            buildGlobalOptionDefinition({
+              name: "param2",
+              description: "param2 description",
+              defaultValue: "default",
+            }),
+            buildGlobalOptionDefinition({
+              name: "param3",
+              description: "param3 description",
+              parameterType: ParameterType.BIGINT,
+              defaultValue: 0n,
+            }),
+          ],
+        },
+      ]);
+
+      setEnvVar("HARDHAT_PARAM3", "5n");
+
+      const globalOptions = resolveGlobalOptions(
+        {
+          param1: "false",
```

Comment:
this is an example of a param that should already be parsed

Reply:
I think a good way to think about this is: first the cli parses the cli arguments, then the hre is initialized, using those arguments, which requires resolving like this.